File: 2024 03/product-of-array-except-self.java
class Solution {
  public int[] productExceptSelf(int[] nums) {
    int n = nums.length, c = -1, total = 1;
    for (int i = 0; i < n; i++) {
      if (nums[i] == 0) {
        if (c == -1) c = i;
        else { return new int[n]; } 
      }
      else { total *= nums[i]; }
    }
    if (c != -1) {
      nums = new int[n];
      nums[c] = total;
      return nums;
    }
    for (int i = 0; i<n; i++) { nums[i] = total/nums[i]; }
    return nums;
  }
}
class product {
  public static void main(String[] args) {
    Solution s = new Solution();
    int[] nums = {1,2,3,4};
    for (int i: s.productExceptSelf(nums)) { System.err.print(i + " "); } System.err.println();
    nums =  new int[]{-1,1,0,-3,3};
    // for (int i: s.productExceptSelf(nums)) { System.err.print(i + " "); } System.err.println();
    nums =  new int[]{-1,1,0,-3,3,0};
    // for (int i: s.productExceptSelf(nums)) { System.err.print(i + " "); } System.err.println();
    
  }
}

File: 2024 07/product-of-array-except-self.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int n = nums.size();
    vector<int> ans(n,1);
    int cur = nums[0];
    for (int i = 1; i < n; i++) { ans[i] = cur; cur *= nums[i]; }
    cur = nums[n-1];
    for (int i = n-2; i >= 0; i--) { ans[i] *= cur; cur *= nums[i]; }
    return ans;
  }
};
int main() {
  Solution s;
  vector<int> nums = {1,2,3,4};
  for (int i : s.productExceptSelf(nums)) { cout << i << " "; }
}